doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2308 @@
+This is not always ideal, since it's sometimes useful to expose a `T` in your type even though you don't own one; we can work around this using the compiler's suggestion: `PhantomData`.
+For more information on how to use it, refer to the type documentation[^110] or the relevant Rustonomicon entry[^111].
+
+
+### Smart Pointers {#smart-pointers}
+
+In Rust, a "smart pointer"[^112] is any type that implements `std::ops::Deref`[^113], the dereference operator[^114].
+`Deref` is defined like this:
+```rust
+trait Deref {
+  type Target;
+  fn deref(&self) -> &Self::Target;
+}
+```
+Types which implement `Deref` can also implement the mutable variant:
+```rust
+trait DerefMut: Deref {
+  fn deref_mut(&mut self) -> &mut Self::Target;
+}
+```
+
+Implementing the `Deref` trait gives a type `T` two features:
+*   It can be dereferenced: `*x` becomes syntax sugar for `*(x.deref())` or `*(x.deref_mut())`, depending on whether the resulting lvalue is assigned to.
+*   It gains auto-deref: if `x.foo` is not a field or method of `T`, then it expands into `x.deref().foo` or `x.deref_mut().foo`, again depending on use.
+
+Furthermore, `deref` and `deref_mut` are called by doing an explicit reborrow: `&*x` and `&mut *x`.
+
+One example of a smart pointer is `ManuallyDrop<T>`.
+Even though this type contains a `T` directly (rather than through a reference), it's still called a "smart pointer", because it can be dereferenced to obtain the `T` inside, and methods of `T` can be called on it.
+As we will see later, the `RefCell<T>` type also produces smart pointers.
+It is not uncommon for generic wrapper types, which restrict access to a value, to be smart pointers.
+
+Note that, because `Target` is an associated type, a type can only dereference to one other type.
+
+While not quite as relevant to smart pointers, the `Index` and `IndexMut` traits are analogous to the `Deref` and `DerefMut` traits, which enable the `x[...]` subscript syntax.
+`Index` looks like this:
+```rust
+trait Index<Idx> {
+  type Output;
+  fn index(&self, index: Idx) -> &Self::Output;
+}
+```
+
+An indexing operation, much like a dereference operation, expands from `x[idx]` into `*(x.index(idx))`.
+Note that indexing _can_ be overloaded, and is a useful example of how this overloading through traits can be useful.
+For example, `<[u8] as Index<usize>>::Output` is `u8`, while `<[u8] as Index<Range>>::Output` is `[u8]`.
+Indexing with a single index produces a single byte, while indexing with a range produces another slice.
+
+
+### Closures {#closures}
+
+Closures (sometimes called "lambda expressions" in other languages) are function literals that capture some portion of their environment, which can be passed into other functions to customize behavior.
+The classic C equivalent is something like `pthread_create`, which takes a function pointer `void* (*)(void*)` and a `void*`, the latter being "captured state" sent across the thread boundary.
```

Comment:
I don't think this is a good analogy -- it's not like pthread_create at all. I think it's safe to say that there is no classic C equivalent, but there are in many other languages.

Reply:
Well, the point I want to get across is that `&dyn Fn` actually behaves that way at the ABI layer. You pass in a function pointer (well, there's an extra vtable indirection...) with some state that's behind a pointer. This is totally a pattern I've seen before; I think qsort() is similar.

I really want an analogy here with a familiar C function, to make it clear that dyn closures are not just a function pointer, but also a pointer to some context.

Reply:
Maybe rather than say "the classic C equivalent" say "the closest analogy in C" and explain how it's different?

Reply:
I wrote some prose. Tell me what you think!

Reply:
Definitely clearer. My one remaining suggestion relates to this text:

"In a similar way, Rust need extra state to execute, except arg becomes part of the start_routine value. As we'll see, Rust has a number of different ABIs for closures, some of which closely resemble what pthread_create does; in some cases, the function pointer can even be inlined."

I'd suggest that you mention that unlike the void* in pthread_create, where this context must be manually constructed, the Rust language automatically constructs it through static analysis.  I think this stuff is obvious to anyone who's used them before but if a pure C person is reading it this will not be clear. 

I'm touching on this in part because closures are so critical in Tock (e.g., grants, MapCell, etc.).

Finally, I think "Rust need" should be "Rust closures need"?

Reply:
Done!